lib/kube/proxy/portforward_test.go

```diff
@@ -0,0 +1,194 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/gravitational/trace"
+	"github.com/stretchr/testify/require"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+
+	testingkubemock "github.com/gravitational/teleport/lib/kube/proxy/testing/kube_server"
+)
+
+func TestPortForwardKubeService(t *testing.T) {
+	const (
+		localPort = 9084
+	)
+	kubeMock, err := testingkubemock.NewKubeAPIMock()
+	require.NoError(t, err)
+	t.Cleanup(func() { kubeMock.Close() })
+
+	// creates a Kubernetes service with a configured cluster pointing to mock api server
+	testCtx := setupTestContext(
+		context.Background(),
+		t,
+		testConfig{
+			clusters: []kubeClusterConfig{{name: kubeCluster, apiEndpoint: kubeMock.URL}},
+		},
+	)
+
+	t.Cleanup(func() { require.NoError(t, testCtx.Close()) })
+
+	// create a user with access to kubernetes (kubernetes_user and kubernetes_groups specified)
+	user, _ := testCtx.createUserAndRole(
+		testCtx.ctx,
+		t,
+		username,
+		roleSpec{
+			name:       roleName,
+			kubeUsers:  roleKubeUsers,
+			kubeGroups: roleKubeGroups,
+		})
+
+	// generate a kube client with user certs for auth
+	_, config := testCtx.genTestKubeClientTLSCert(
+		t,
+		user.GetName(),
+		kubeCluster,
+	)
+	require.NoError(t, err)
+
+	type args struct {
+		portforwardBuilder func(*testing.T, portForwardRequestConfig) forwardPorts
+	}
+	tests := []struct {
+		name string
+		args args
+	}{
+		{
+			name: "SPDY protocol",
+			args: args{
+				portforwardBuilder: spdyPortForward,
```

Comment:
Is it viable to test using public methods, or maybe test through using Teleport itself?

Reply:
This test already launches an instance of the Teleport Kubernetes Service, configures the user, and makes a request through the API to carry out a port-forward.